website/docs/language/expressions/custom-validation-rules.mdx

```diff
@@ -0,0 +1,316 @@
+  # the EC2 instance declared by aws_instance.example, using the exported ID.
+
+  filter {
+    name = "volume-id"
+    values = [aws_instance.example.root_block_device.volume_id]
+  }
+}
+
+output "api_base_url" {
+  value = "https://${aws_instance.example.private_dns}:8433/"
+
+  # An output value with a precondition can check the object that the
+  # output value is describing to make sure it meets expectations before
+  # any caller of this module can use it.
+  precondition {
+    condition     = data.aws_ebs_volume.example.encrypted
+    error_message = "The server's root volume is not encrypted."
+  }
+}
+```
+
+The preconditions and postconditions declare the following assumptions and guarantees.
+
+- **The AMI ID must refer to an AMI that exists and that has been tagged with "nomad-server".** This would detect if the caller accidentally provided an AMI intended for some other system component, which might otherwise be detected only after booting the EC2 instance and noticing that the expected network service is not running.
+
+- **The AMI ID must refer to an AMI that contains an operating system for the
+  `x86_64` architecture.** This would detect if the caller accidentally built an AMI for a different
+    architecture, which may not be able to run the software this virtual machine is intended to host.
+
+- **The EC2 instance must be allocated a private DNS hostname.** In Amazon Web Services, EC2 instances are assigned private DNS hostnames only if they belong to a virtual network configured in a certain way. This would detect if the selected virtual network is not configured correctly, giving explicit feedback to prompt the user to debug the network settings.
+
+- **The EC2 instance will have an encrypted root volume.** This ensures that the root volume is encrypted even though the software running in this EC2 instance would probably still operate as expected on an unencrypted volume. This lets Terraform produce an error immediately, before any other components rely on the insecurely-configured component.
+
+## Condition Expressions
+
+The variable `validation` block and the `precondition` and `postcondition` blocks all require an argument named `condition`, whose value is a boolean expression which should return `true` if the intended assumption holds or `false` if it does not.
+
+Condition expressions have the following requirements.
+
+- For variable `validation` blocks, the expression can refer only to the variable that the condition applies to and must not produce errors.
+- For `precondition` and `postcondition` blocks, the expression can refer to any other objects in the same module, as long as the references don't create any cyclic dependencies. Resource postconditions can additionally refer to attributes of each instance of the resource where they are configured, using the special symbol `self`. For example, `self.private_dns` refers to the `private_dns` attribute of each instance of the containing resource.
```

Comment:
I think we should document somewhere the availability of `count.index`, `each.key`, and `each.value` for expanded resources. This is useful for writing conditions which refer to values in peer resource instances which share the same `count` or `for_each` expression.

Reply:
Also relevant to this, I think, is the fact that we will evaluate preconditions _after_ we've evaluated the `count` or `for_each` argument, if any, so that Terraform can evaluate the preconditions separately for each instance and thus make `each.key`, `each.value`, and  `count.index` available to those conditions. The important implication of that is that the `count` and `for_each` expressions are _not_ guarded by the preconditions and so must be valid even when the precondition doesn't hold.

I'm not if _this_ is the ideal place for that or if it would be better to include it as another item in this list from earlier in the document:

https://github.com/hashicorp/terraform/blob/040985f6e10b245a1f870f9c8c0bff7de8de3984/website/docs/language/expressions/custom-validation-rules.mdx#L84-L88

(Tangentially, this does make me consider a possibly later refinement of this feature where we could potentially evaluate the preconditions earlier if the condition and error message expressions contain no references to `count.index`, `each.key`, or `each.value`, but perhaps that'd be "too magical" to be understandable to a non-expert reader of an existing configuration.)

Reply:
gotcha - thank you both :)  I think we should do two things:
- Add a bullet or a sentence onto an existing bullet about how preconditions are evaluated after count and for_each. I'll do that today and tag you for review!
- Add a heading under "Condition Expressions" about how these things are available. We could provide a bit of context and show an example of why it's useful, as we do for some of the other ones. @alisdair would you be open to just giving me a draft of what we'd want that text to be? Then I can add it in and refine the language. 

Thank you both!

Reply:
Sorry this took so long, I got into a deep rabbit hole trying to find a reasonable use of chained expanded resources. I hope this is helpful in some way!

Here are a couple of suggested sub-sections under "Condition Expressions". Once I had written the example about `each` and `count`, it seemed odd not to call out the (probably more common) use of `self` at the same level. I realize it's currently part of a bullet point very nearby, but giving the two expression values parity felt like a reasonable idea to propose.

(Using `<pre>` because I can't have Markdown code blocks inside a Markdown code block 😵‍💫)

<pre>
### `self` object

You can use the `self` object in postcondition blocks to refer to attributes of the instance under evaluation.

```hcl
resource "aws_instance" "example" {
  instance_type = "t2.micro"
  ami           = "ami-abc123"

  lifecycle {
    postcondition {
      condition     = self.instance_state == "running"
      error_message = "EC2 instance must be running."
    }
  }
}
```

### `each` and `count` objects

In blocks where `for_each` or `count` are set, corresponding `each` and `count` objects are available for use in conditions. These can be used to refer to other resources which are expanded in a chain.

```hcl
variable "vpc_cidrs" {
  type = set(string)
}

data "aws_vpc" "example" {
  for_each = var.vpc_cidrs

  filter {
    name   = "cidr"
    values = [each.key]
  }
}

resource "aws_internet_gateway" "example" {
  for_each = aws_vpc.example
  vpc_id = each.value.id

  lifecycle {
    precondition {
      condition     = aws_vpc.example[each.key].state == "available"
      error_message = "VPC ${each.key} must be available."
    }
  }
}
```
</pre>

Reply:
awesome! I'm going to add this in - thank you!